SelfOrganisingList.tpp

```diff
@@ -8,20 +8,77 @@ SelfOrganisingList<T>::SelfOrganisingList(int _size) : size(_size), count(0){
 
 template<typename T>
 void SelfOrganisingList<T>::add(T value) {
-    if (count < size-1){
-        auto *tmp = new Node<T>(value);
-        tab[count++] = *tmp;
+    if (count >= size) {
+
+        return; //PELNA LISTA
+    }
+
+    if (indexOf(value) != -1) {
+
+        return; //ELEMENT JUZ ISTNIEJE
     }
+
+    tab[count++] = Node<T>(value);
+}
+
+template<typename T>
+void SelfOrganisingList<T>::addAt(T value, int idx) {
+    if (count >= size) {
+
+        return; //PELNA LISTA
+    }
+
+    if (idx < 0 || idx > count) {
+
+        return; //NIEWLASCIWY INDEX
+    }
+
+    if (indexOf(value) != -1) {
+
+        return; //ELEMENT JUZ ISTNIEJE
+    }
+
+    for (int i = count; i > idx; i--) {
+        tab[i] = tab[i - 1];
+    }
+
+    tab[idx] = Node<T>(value);
+    count++;
 }
 
 template<typename T>
 void SelfOrganisingList<T>::remove(T value) {
-    count--;
+    for (int i = 0; i < count; i++) {
+        if (tab[i].getValue() == value) {
+            for (int j = i; j < count - 1; j++) {
+                tab[j] = tab[j + 1];
+            }
+            count--;
+            tab[count] = Node<T>();
```

Comment:
you could change that two lines into tab[--count] = ...
also - when removed from array, we should also delete it from memory i think
another thing - why new empty node is created after whole process?